### PR TITLE
#9263 geoprocessing tools

### PR DESCRIPTION
**Pull request by MV88**
## Description

this PR implements the buffer and intersection tool. for more details see the issue

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

- #9263

**What is the new behavior?**

- you can use buffer on one layer and intersection on two layers

see issue description for more details

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] Comments in line
- [ ] Unit tests
- [x] Adding it to the layouts is not a good practice, mapstore should not depend by its plugin. Can you ask @allyoucanmap a more efficient way to properly set the layout for new plugins, if any?
- [ ] the plugin starts in opened state. It shouldn't. At list it should be optional
- [ ] Enable highlight of selected features strings --> say simply "Highlight" and put it inline. 
image
- [ ] Intersection: When doing this selection 

image

I have this error in console without any feedkback to the user.
```
geoProcessingTools.js:479 WPSExecuteError: Failed to retrieve value for input first feature collection
    at handleExecuteResponseXMLObject (execute.js:119:11)
    at MergeMapSubscriber.eval [as project] (execute.js:286:9)
    at MergeMapSubscriber._tryNext (mergeMap.js:120:27)
    at MergeMapSubscriber._next (mergeMap.js:110:18)
    at Subscriber.next (Subscriber.js:89:18)
    at SwitchMapSubscriber.notifyNext (switchMap.js:124:30)
    at InnerSubscriber._next (InnerSubscriber.js:23:21)
    at Subscriber.next (Subscriber.js:89:18)
    at MapSubscriber._next (map.js:83:26)
    at Subscriber.next (Subscriber.js:89:18)
    at OuterSubscriber.notifyNext (OuterSubscriber.js:19:26)
    at InnerSubscriber._next (InnerSubscriber.js:23:21)
    at Subscriber.next (Subscriber.js:89:18)
    at eval (subscribeToResult.js:36:29)
```

- [ ] Buffer: I did a buffer, the layer is added to the TOC but I can not see anything. Even changing he style do not produce any result
image

File: web/client/plugins/GeoProcessingTools.jsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * @prop {string} cfg.capStyle (ONLY BUFFER) Style for the buffer end caps. Values are: Round - rounded ends (default), Flat - flat ends; Square - square ends, default is unset
+ * @prop {string} cfg.selectedTool the values are "buffer", "intersection", default is "buffer"
+ * @prop {string} cfg.firstAttributesToRetain (ONLY INTERSECTION) First feature collection attribute to include
+ * @prop {string} cfg.secondAttributesToRetain (ONLY INTERSECTION) Second feature collection attribute to include
+ * @prop {string} cfg.intersectionMode (ONLY INTERSECTION) Specifies geometry computed for intersecting features. INTERSECTION (default) computes the spatial intersection of the inputs. FIRST copies geometry A. SECOND copies geometry B.
+ * @prop {boolean} cfg.percentagesEnabled (ONLY INTERSECTION) Indicates whether to output feature area percentages (attributes percentageA and percentageB)
+ * @prop {boolean} cfg.areasEnabled (ONLY INTERSECTION) Indicates whether to output feature areas (attributes areaA and areaB)
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/plugins/GeoProcessingTools.jsx:38]: firstAttributesToRetain is the name used in the wps process so i aligned to it.

moving cfg into sections makes sense but it forces me to refactor the INIT_PLUGIN case in the reducer

image

File: web/client/themes/default/less/geo-processing-tools.less [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+// **************
+// Theme
+// **************
+#ms-components-theme(@theme-vars) {
+    // #GeoProcessingTools-root {
+
+    // }
+}
+
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/themes/default/less/geo-processing-tools.less:17]: usually we keep it empty

**Comment by MV88** [comment]
> he plugin starts in opened state. It shouldn't. At list it should be optional

yeah sure it was done for testing, will be removed

>  Enable highlight of selected features strings --> say simply "Highlight" and put it inline.

this highlight works only for selected features not for the entire layer, one user can misinterpret. i do not understand why we have to be vague instead of specific

@offtherailz do you have any suggestion on the icon?

File: web/client/utils/WFSLayerUtils.js [hunk 3]
```diff
@@ -60,3 +59,27 @@ export const extractGeometryAttributeName = describeFeatureType => {
...
+ *     }]
+ * }]
+ */
+export const extractFirstNonGeometryProp = describeFeatureType => {
+    const properties = get(describeFeatureType, "featureTypes[0].properties") || [];
+    return properties && head(properties
+        .filter(elem => elem.type.indexOf("gml:") !== 0)
+        .map(elem => elem.name)
+    );
+};
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/utils/WFSLayerUtils.js:85]: should be called **findNonGeometryProperty**

**Comment by MV88** [comment]
- in my case the buffer works

image

- for the error of the intersection this is the request sent, and it seems ok on a first look.
  - I will add a notification
  - we have to check if there is a problem in geoserver maybe

```
curl --location 'https://gs-stable.geosolutionsgroup.com/geoserver/wps?service=WPS&version=1.0.0&REQUEST=Execute' \
--header 'authority: gs-stable.geosolutionsgroup.com' \
--header 'accept: application/xml, application/json' \
--header 'accept-language: en-US,en;q=0.9' \
--header 'cache-control: no-cache' \
--header 'content-type: application/xml' \
--header 'origin: http://localhost:8081' \
--header 'pragma: no-cache' \
--header 'referer: http://localhost:8081/' \
--header 'sec-ch-ua: "Not/A)Brand";v="99", "Google Chrome";v="115", "Chromium";v="115"' \
--header 'sec-ch-ua-mobile: ?0' \
--header 'sec-ch-ua-platform: "Windows"' \
--header 'sec-fetch-dest: empty' \
--header 'sec-fetch-mode: cors' \
--header 'sec-fetch-site: cross-site' \
--header 'user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36' \
--data '<?xml version="1.0" encoding="UTF-8"?>
<wps:Execute version="1.0.0" service="WPS" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.opengis.net/wps/1.0.0" xmlns:wfs="http://www.opengis.net/wfs" xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:gml="http://www.opengis.net/gml" xmlns:ogc="http://www.opengis.net/ogc" xmlns:wcs="http://www.opengis.net/wcs/1.1.1" xmlns:dwn="http://geoserver.org/wps/download" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsAll.xsd">
    <ows:Identifier>gs:IntersectionFeatureCollection</ows:Identifier>
    <wps:DataInputs>
        <wps:Input>
            <ows:Identifier>first feature collection</ows:Identifier>
            <wps:Data>
                <wps:ComplexData mimeType="application/json">
                    <![CDATA[{"type":"FeatureCollection","features":[{"type":"Feature","geometry":{"type":"MultiPolygon","coordinates":[[[[-88.0716,37.511],[-88.0879,37.4763],[-88.3117,37.4429],[-88.3592,37.4093],[-88.4199,37.4203],[-88.4676,37.4008],[-88.5113,37.2969],[-88.5014,37.2578],[-88.4507,37.2057],[-88.4225,37.1569],[-88.4505,37.0987],[-88.4768,37.0721],[-88.4907,37.0682],[-88.5173,37.0648],[-88.5593,37.0728],[-88.6142,37.109],[-88.6884,37.1354],[-88.7391,37.1412],[-88.7465,37.1521],[-88.8633,37.2022],[-88.9325,37.2184],[-88.9932,37.22],[-89.065,37.1859],[-89.1168,37.1121],[-89.1463,37.0932],[-89.1695,37.0642],[-89.1743,37.0257],[-89.1502,36.9984],[-89.1299,36.9881],[-89.1935,36.9868],[-89.2101,37.029],[-89.2377,37.0417],[-89.2641,37.0871],[-89.2842,37.0912],[-89.3033,37.0854],[-89.3097,37.0609],[-89.2642,37.0277],[-89.262,37.0087],[-89.2828,36.9992],[-89.311,37.0097],[-89.3829,37.0492],[-89.38,37.0991],[-89.4238,37.1372],[-89.4405,37.1653],[-89.4682,37.2243],[-89.4653,37.2537],[-89.4896,37.256],[-89.5139,37.2764],[-89.5139,37.305],[-89.5006,37.3294],[-89.4687,37.3394],[-89.4357,37.3557],[-89.4276,37.411],[-89.4536,37.4532],[-89.4948,37.4917],[-89.525,37.572],[-89.5134,37.6159],[-89.5192,37.6504],[-89.5134,37.6798],[-89.5215,37.6948],[-89.5814,37.7061],[-89.6665,37.7455],[-89.6759,37.784],[-89.6911,37.8048],[-89.7284,37.841],[-89.8517,37.9051],[-89.861,37.9055],[-89.8668,37.8919],[-89.9006,37.8759],[-89.9379,37.878],[-89.9789,37.9119],[-89.9582,37.9636],[-90.0108,37.9693],[-90.0419,37.9932],[-90.1193,38.0323],[-90.1347,38.054],[-90.2075,38.0889],[-90.2541,38.1222],[-90.2896,38.1668],[-90.3367,38.1887],[-90.3648,38.2343],[-90.3693,38.3236],[-90.3587,38.3653],[-90.3396,38.3908],[-90.3018,38.4274],[-90.2658,38.5187],[-90.2612,38.5328],[-90.2409,38.5628],[-90.1837,38.6103],[-90.1836,38.6588],[-90.2022,38.7004],[-90.1966,38.724],[-90.1634,38.7731],[-90.1352,38.7855],[-90.1217,38.8005],[-90.1131,38.8305],[-90.1328,38.853],[-90.2439,38.9145],[-90.2789,38.9247],[-90.3197,38.9249],[-90.4131,38.9623],[-90.4698,38.9592],[-90.5304,38.8916],[-90.5703,38.8713],[-90.6272,38.8808],[-90.6689,38.9353],[-90.7061,39.0378],[-90.7076,39.0582],[-90.6904,39.0937],[-90.7167,39.1442],[-90.7182,39.1959],[-90.7323,39.2247],[-90.7381,39.2478],[-90.7793,39.2968],[-90.8505,39.3505],[-90.9479,39.4006],[-91.0363,39.4444],[-91.0644,39.474],[-91.0936,39.5289],[-91.1562,39.5526],[-91.2032,39.6],[-91.3177,39.6859],[-91.3671,39.7246],[-91.3734,39.7613],[-91.3817,39.8038],[-91.4492,39.863],[-91.451,39.8852],[-91.4341,39.9018],[-91.4304,39.9218],[-91.4472,39.9461],[-91.4873,40.0058],[-91.504,40.0667],[-91.5161,40.1345],[-91.5065,40.2005],[-91.4989,40.2514],[-91.4867,40.3096],[-91.4486,40.3719],[-91.4188,40.3869],[-91.3858,40.3924],[-91.3728,40.403],[-91.3854,40.4472],[-91.3748,40.5037],[-91.3821,40.5285],[-91.4129,40.548],[-91.4111,40.573],[-91.3756,40.6034],[-91.2621,40.6395],[-91.2149,40.6438],[-91.1625,40.6563],[-91.1292,40.6821],[-91.12,40.7054],[-91.0928,40.7615],[-91.0889,40.8337],[-91.0492,40.8796],[-90.9833,40.9239],[-90.9607,40.9505],[-90.9547,41.0704],[-90.9578,41.1044],[-90.9903,41.1444],[-91.0183,41.1658],[-91.0563,41.1763],[-91.1015,41.2315],[-91.1023,41.2678],[-91.0733,41.3349],[-91.0558,41.4014],[-91.0275,41.4235],[-91.0007,41.4311],[-90.9497,41.4212],[-90.8441,41.4446],[-90.7799,41.4498],[-90.7082,41.4501],[-90.6588,41.4623],[-90.6007,41.5096],[-90.5408,41.526],[-90.455,41.5275],[-90.435,41.5436],[-90.423,41.5673],[-90.3484,41.5868],[-90.3393,41.6028],[-90.3411,41.6491],[-90.326,41.7227],[-90.3049,41.7565],[-90.2553,41.7817],[-90.1958,41.8061],[-90.1545,41.9308],[-90.1427,41.984],[-90.1505,42.0334],[-90.1681,42.061],[-90.1666,42.1037],[-90.1761,42.1205],[-90.1916,42.1227],[-90.2309,42.1597],[-90.3236,42.1973],[-90.3677,42.2102],[-90.4072,42.2426],[-90.418,42.2639],[-90.4277,42.3406],[-90.4416,42.3601],[-90.491,42.3888],[-90.5636,42.4218],[-90.6058,42.4606],[-90.6483,42.4756],[-90.6518,42.4947],[-90.6383,42.5094],[-90.42,42.5084],[-89.9236,42.5041],[-89.8346,42.5035],[-89.4005,42.4975],[-89.3594,42.4979],[-88.9391,42.4909],[-88.765,42.4909],[-88.7065,42.4897],[-88.2979,42.492],[-88.1947,42.4896],[-87.7973,42.4891],[-87.8369,42.3142],[-87.7602,42.1565],[-87.6705,42.0598],[-87.6126,41.8473],[-87.5299,41.7236],[-87.5326,41.4697],[-87.5324,41.3013],[-87.5317,41.1738],[-87.532,41.0099],[-87.5327,40.7454],[-87.5372,40.4946],[-87.5357,40.4832],[-87.5353,40.1662],[-87.5358,39.8873],[-87.5356,39.6093],[-87.5386,39.4774],[-87.5402,39.3505],[-87.5977,39.3383],[-87.6252,39.3074],[-87.6106,39.2977],[-87.6158,39.2814],[-87.6069,39.2582],[-87.5846,39.2488],[-87.5886,39.2085],[-87.5942,39.1981],[-87.6079,39.1961],[-87.6443,39.1685],[-87.6703,39.1467],[-87.6595,39.1307],[-87.6623,39.1135],[-87.6317,39.1039],[-87.6309,39.089],[-87.612,39.0846],[-87.5853,39.0624],[-87.5817,38.9957],[-87.5919,38.9941],[-87.5479,38.9771],[-87.5335,38.9637],[-87.5302,38.9319],[-87.5392,38.9049],[-87.5591,38.8698],[-87.5505,38.8579],[-87.5079,38.7956],[-87.519,38.7767],[-87.508,38.7697],[-87.5083,38.7366],[-87.5439,38.686],[-87.5885,38.6722],[-87.6252,38.6428],[-87.6286,38.6229],[-87.6198,38.5992],[-87.6406,38.5932],[-87.6529,38.5739],[-87.6729,38.5474],[-87.6514,38.5154],[-87.6535,38.5004],[-87.6799,38.504],[-87.6928,38.4815],[-87.7561,38.4661],[-87.7587,38.4571],[-87.739,38.4455],[-87.7484,38.418],[-87.784,38.3781],[-87.8345,38.3525],[-87.8501,38.2861],[-87.863,38.2854],[-87.874,38.3168],[-87.8834,38.3156],[-87.8885,38.3007],[-87.9141,38.281],[-87.9137,38.3023],[-87.9259,38.3048],[-87.98,38.2411],[-87.986,38.2348],[-87.9779,38.2007],[-87.9323,38.1711],[-87.932,38.1575],[-87.9506,38.1369],[-87.9735,38.1318],[-88.0185,38.1033],[-88.0123,38.0923],[-87.9649,38.0967],[-87.9753,38.0733],[-88.0347,38.0541],[-88.0431,38.0451],[-88.0415,38.0383],[-88.0217,38.0335],[-88.0292,38.0082],[-88.0217,37.9751],[-88.0425,37.9563],[-88.0418,37.9345],[-88.0646,37.9298],[-88.0789,37.944],[-88.084,37.9237],[-88.0304,37.9176],[-88.0266,37.9058],[-88.0449,37.896],[-88.1001,37.9062],[-88.1015,37.8953],[-88.0757,37.8678],[-88.0342,37.8437],[-88.0421,37.8275],[-88.0893,37.8312],[-88.086,37.8176],[-88.0356,37.8057],[-88.0725,37.7354],[-88.1336,37.7007],[-88.1594,37.6607],[-88.1576,37.6285],[-88.1342,37.5836],[-88.0716,37.511]]]]}}]}]]>
                </wps:ComplexData>
            </wps:Data>
        </wps:Input>
        <wps:Input>
            <ows:Identifier>second feature collection</ows:Identifier>
            <wps:Data>
                <wps:ComplexData mimeType="application/json">
                    <![CDATA[{"type":"FeatureCollection","features":[{"type":"Feature","geometry":{"type":"MultiPolygon","coordinates":[[[[-77.0082,38.9666],[-76.9112,38.89],[-77.0454,38.7881],[-77.0352,38.8139],[-77.0452,38.8294],[-77.0404,38.8384],[-77.0391,38.8624],[-77.0679,38.8861],[-77.0789,38.9156],[-77.1226,38.9321],[-77.0424,38.9934],[-77.0082,38.9666]]]]}}]}]]>
                </wps:ComplexData>
            </wps:Data>
        </wps:Input>
        <wps:Input>
            <ows:Identifier>intersectionMode</ows:Identifier>
            <wps:Data>
                <wps:LiteralData>INTERSECTION</wps:LiteralData>
            </wps:Data>
        </wps:Input>
        <wps:Input>
            <ows:Identifier>percentagesEnabled</ows:Identifier>
            <wps:Data>
                <wps:LiteralData>false</wps:LiteralData>
            </wps:Data>
        </wps:Input>
        <wps:Input>
            <ows:Identifier>areasEnabled</ows:Identifier>
            <wps:Data>
                <wps:LiteralData>false</wps:LiteralData>
            </wps:Data>
        </wps:Input>
    </wps:DataInputs>
    <wps:ResponseForm>
        <wps:RawDataOutput mimeType="application/json">
            <ows:Identifier>result</ows:Identifier>
        </wps:RawDataOutput>
    </wps:ResponseForm>
</wps:Execute>'
```

with this stack trace in geoserver 
```

28 Jul 16:30:58 ERROR  [wps.executor] - Process execution failed
org.geoserver.wps.WPSException: Failed to retrieve value for input first feature collection
	at org.geoserver.wps.executor.LazyInputMap.parseInputs(LazyInputMap.java:98)
	at org.geoserver.wps.executor.LazyInputMap.get(LazyInputMap.java:50)
	at org.geotools.process.factory.AnnotationDrivenProcessFactory$InvokeMethodProcess.buildProcessArguments(AnnotationDrivenProcessFactory.java:748)
	at org.geotools.process.factory.AnnotationDrivenProcessFactory$InvokeMethodProcess.execute(AnnotationDrivenProcessFactory.java:628)
	at org.geoserver.wps.executor.ProcessStartupFilter$ProcessStartupWrapper.execute(ProcessStartupFilter.java:50)
	at org.geoserver.wps.executor.DefaultProcessManager$ProcessCallable.call(DefaultProcessManager.java:222)
	at org.geoserver.wps.executor.DefaultProcessManager$ProcessCallable.call(DefaultProcessManager.java:191)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at org.geotools.data.geojson.GeoJSONReader.getBuilder(GeoJSONReader.java:406)
	at org.geotools.data.geojson.GeoJSONReader.getNextFeature(GeoJSONReader.java:283)
	at org.geotools.data.geojson.GeoJSONReader.getFeatures(GeoJSONReader.java:235)
	at org.geoserver.wps.ppio.GeoJSONPPIO$FeatureCollections.decode(GeoJSONPPIO.java:93)
	at org.geoserver.wps.ppio.CDataPPIO.decode(CDataPPIO.java:21)
	at org.geoserver.wps.executor.SimpleInputProvider.getValueInternal(SimpleInputProvider.java:81)
	at org.geoserver.wps.executor.AbstractInputProvider.getValue(AbstractInputProvider.java:103)
	at org.geoserver.wps.executor.LazyInputMap.parseInputs(LazyInputMap.java:90)
	... 10 more
```

**Comment by MV88** [comment]
closed in favor of this 
https://github.com/geosolutions-it/MapStore2/pull/9316/commits

**Pull request closed without merging**